IA/IA_upload.py

```diff
@@ -0,0 +1,104 @@
+import os
+import argparse
+import requests
+import boto
+from boto.s3.connection import OrdinaryCallingFormat
+from boto.s3.multipart import MultiPartUpload
+import asyncio
+import xmltodict
+from io import BytesIO
+from settings import (
+    CHUNK_SIZE,
+    OSF_COLLECTION_NAME,
+    IA_ACCESS_KEY,
+    IA_SECRET_KEY,
+    IA_URL
+)
+
+HERE = os.path.dirname(os.path.abspath(__file__))
+
+
+def mp_from_ids(mp_id: str, mp_keyname: str, bucket: str) -> MultiPartUpload:
+    mp = MultiPartUpload(bucket)
+    mp.key_name = mp_keyname
+    mp.id = mp_id
+    return mp
+
+
+async def gather_and_upload(bucket_name: str, parent: str):
+    '''
+    This script traverses through a directory uploading everything in it to Internet Archive.
+    '''
+
+    tasks = []
+
+    for root, dirs, files in os.walk(parent):
+        for file in files:
+            path = os.path.join(root, file)
+            with open(path, 'rb') as fp:
+                data = fp.read()
+                size = len(data)
+                if size > CHUNK_SIZE:
+                    tasks.append(chunked_upload(bucket_name, path, data))
+                else:
+                    tasks.append(upload(bucket_name, path, data))
+
+    await asyncio.gather(*tasks)
+
+
+async def upload(bucket_name: str, filename: str, file_content: bytes):
+    headers = {
+        'authorization': 'LOW {}:{}'.format(IA_ACCESS_KEY, IA_SECRET_KEY),
+        'x-amz-auto-make-bucket': '1',
+        'Content-Type': 'application/octet-stream',
+        'x-archive-meta01-collection': OSF_COLLECTION_NAME,
+    }
+    url = f'{IA_URL}/{bucket_name}/{filename}'
+    resp = requests.put(url, headers=headers, data=file_content, stream=True)
+
+    if resp.status_code != 200:
```

Comment:
The only success code they use is 200? Will this automatically handle the [likely 307 response](https://archive.org/services/docs/api/ias3.html#how-this-is-different-from-normal-s3) from IA? Also, what about [the 503 slow down](https://archive.org/services/docs/api/ias3.html#use-limits) responses? Is there a graceful way of dealing with those?

Reply:
for the 307s, `requests` handles that I believe and will automatically redirect and return the first non-300. Also I added a `put_with_retry` function for retries.